### ISSUE TITLE
V6 mudflow doesn’t work as intended

### ISSUE THREAD

**Issue by numberZero** [minetest/minetest#8784]
##### Minetest version
`master`

##### Summary
https://github.com/minetest/minetest/blob/2f879e8bbd8329e4c796dc1f7eeea9bdc3be2b3d/src/mapgen/mapgen_v6.cpp#L779-L787
This code was probably intended to run several iterations mud flow in different directions. But in fact, 0th and 2nd iterations are only executed for the last position.

https://github.com/minetest/minetest/blob/2f879e8bbd8329e4c796dc1f7eeea9bdc3be2b3d/src/mapgen/mapgen_v6.cpp#L796-L798
That should be single loop. Also `i` needs to be changed on each iteration, otherwise it may go out of sync with `y`. It will actually, should the loop reach dirt on non-dirt, due to `continue` at line 831.

https://github.com/minetest/minetest/blob/2f879e8bbd8329e4c796dc1f7eeea9bdc3be2b3d/src/mapgen/mapgen_v6.cpp#L824-L827
This check is wrong. Either the area has enough margin below the chunk generated, and the check always passes, or it hasn’t, so that `y` can touch the area bottom, but then `add_y` will make `i2` index node out of column but not out of area, generally.

**Comment by paramat** [comment]
> But in fact, 0th and 2nd iterations are only executed for the last position.

By 'last position' do you mean x = mudflow_maxpos, z = mudflow_maxpos?
If so i can't see how, the co-ordinate inversion looks to work correctly on iterations 0 and 2.

Do you mean, since coordinates are inverted every 2nd iteration, we should have 4 iterations instead of 3?

Thanks for this, i'll investigate.

**Comment by numberZero** [comment]
> By 'last position' do you mean x = mudflow_maxpos, z = mudflow_maxpos?
> If so i can't see how

Yes. Running the loop line by line:
780. (loop initialization) `k` is set to `0`
781. (loop initialization) `z` is set to `mudflow_minpos`
782. (loop initialization) `x` is set to `mudflow_minpos`
785. `x` is changed to `mudflow_maxpos`
786. `z` is changed to `mudflow_maxpos`
... (loop body)
782. (loop step) `x` is increased by `1`, becoming `mudflow_maxpos + 1`
782. (loop condition) `x` is now greater than `mudflow_maxpos`, so this loop ends
781. (loop step) `z` is increased by `1`, becoming `mudflow_maxpos + 1`
781. (loop condition) `z` is now greater than `mudflow_maxpos`, so this loop ends too
780. (loop step) `k` is increased by `1`, becoming `1`
... (this iteration works normally)

**Comment by paramat** [comment]
Ahhh i see, thanks.
>  calculate p2d differently on even and odd iterations.

I agree, will do this.

**Comment by paramat** [comment]
I'll make a branch, confirm the bug, fix it. Then we can iterate mudflow twice, since it is currently effectively iterated only once, then one iteration will have reversed coordinates.

**Comment by paramat** [comment]
Confirmed:
```
k 0
z -15
x-15
k 1
z -15
x-15 x-14 x-13 x-12 x-11 x-10 x-9 x-8 x-7 x-6 x-5 x-4 x-3 x-2 x-1 x0 x1 x2 x3 x4 .....
z -14
x-15 x-14 x-13 x-12 x-11 x-10 x-9 x-8 x-7 x-6 x-5 x-4 x-3 x-2 x-1 x0 x1 x2 x3 x4 .....
.....
.....
z 94
x-15 x-14 x-13 x-12 x-11 x-10 x-9 x-8 x-7 x-6 x-5 x-4 x-3 x-2 x-1 x0 x1 x2 x3 x4 .....
k 2
z -15
x-15 
```

**Comment by paramat** [comment]
PR #8795

**Comment by paramat** [comment]
> This check is wrong.

Agreed, y only ever goes down to node_min.Y (mapchunk base) so 1 below this is still within the voxelmanip (which extends down to node_min.Y - 16).
Check is unnecessary. I'll remove it.

**Comment by paramat** [comment]
> should the loop reach dirt on non-dirt, due to continue at line 831.

I suspect that `continue` should be a `break`. If node below is non-dirt (stone) there's no point starting a new search for dirt down through the column, better to move to new column.

**Comment by paramat** [comment]
```
				// Check that upper is walkable. Cancel
				// dropping if upper keeps it in place.
				u32 i3 = i;
				VoxelArea::add_y(em, i3, 1);
				MapNode *n3 = NULL;

				if (vm->m_area.contains(i3)) {
					n3 = &vm->m_data[i3];
					if (ndef->get(*n3).walkable)
						continue;
				}
```
Likewise this 'out of area' check is unnecessary, as y is never higher than node_max.Y, the voxelmanip area extends 16 nodes above that.

**Comment by paramat** [comment]
```
					// Check that under side is air
					VoxelArea::add_y(em, i2, -1);
					if (!vm->m_area.contains(i2))
						continue;
					n2 = &vm->m_data[i2];
```
Here too.

**Comment by paramat** [comment]
Another PR #8805

**Comment by paramat** [comment]
> That should be single loop

It is tempting, but it seems intentional, i am concerned it might not work identically rewritten. 
I prefer to leave this as it is for now, as it is not a problem.

**Comment by numberZero** [comment]
> it seems intentional

It seems to be a quick-and-dirty hack. The author hasn’t even bothered with fixing indentation. But you’re right care is necessary to merge these conditions, due to many `break`s and `continue`s in the code. That should be easier after your fixes, though.

**Comment by paramat** [comment]
I'm fairly sure a normal loop will be ok, but not totally sure, would prefer to leave that to a follow-up PR.
Good point about indentation, will fix that in the PR.

**Comment by paramat** [comment]
Ugh, just realised that checking for index being within area cannot act as a reliable check for position within area, so the below will need to be rewritten:
https://github.com/minetest/minetest/blob/ba8fb774c1abe77c404b578929c6d2b657bf1436/src/mapgen/mapgen_v6.cpp#L873-L884
